core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala

```diff
@@ -47,43 +47,48 @@ class SnapshotService(protected val ctx: RawlsRequestContext,
     with WorkspaceSupport
     with LazyLogging {
 
-  def createSnapshot(workspaceName: WorkspaceName, snapshot: NamedDataRepoSnapshot): Future[DataRepoSnapshotResource] =
+  def createSnapshot(workspaceName: WorkspaceName,
+                     snapshotIdentifiers: NamedDataRepoSnapshot
+  ): Future[DataRepoSnapshotResource] =
     getV2WorkspaceContextAndPermissions(workspaceName,
                                         SamWorkspaceActions.write,
                                         Some(WorkspaceAttributeSpecs(all = false))
-    ).flatMap { workspaceContext =>
-      val wsid = workspaceContext.workspaceIdAsUUID // to avoid UUID parsing multiple times
-      // create the stub workspace in WSM if it does not already exist
-      if (!workspaceStubExists(wsid, ctx)) {
-        workspaceManagerDAO.createWorkspace(wsid, ctx)
+    ).flatMap { rawlsWorkspace =>
+      val wsid = rawlsWorkspace.workspaceIdAsUUID // to avoid UUID parsing multiple times
+      val snapshot = dataRepoDAO.getSnapshot(snapshotIdentifiers.snapshotId, ctx.userInfo.accessToken)
+      val validator = new SnapshotReferenceCreationValidator(rawlsWorkspace, snapshot)
+      val aggregatedWorkspaceService = new AggregatedWorkspaceService(workspaceManagerDAO)
+
+      // prevent disallowed access across workspace or dataset protection boundaries
+      validator.validateProtectedStatus()
+
+      Try(aggregatedWorkspaceService.getAggregatedWorkspace(rawlsWorkspace, ctx)) match {
+        case Success(wsmWorkspace) =>
+          // if there's an existing WSM workspace, make sure its platform is compatible
+          // with that of the snapshot's dataset.
+          validator.validatePlatforms(wsmWorkspace.getCloudPlatform)
+        case Failure(_) =>
+          // Pre-emptively fail (and prevent creation of a workspace stub here if snapshot is Azure). We can remove this
+          // check once we support Azure-based snapshots by reference.
+          validator.validateSnapshotPlatform()
```

Comment:
if `aggregatedWorkspaceService.getAggregatedWorkspace(rawlsWorkspace, ctx)`, doesn't it indicate a pretty serious failure, like a totally missing workspace? Is there a point to validating the snapshot in such a case, since import would fail no matter what?

Reply:
Maybe this gets clearer/sorted out once I understand the decision logic for when it's appropriate to call `createWorkspace`, but I'll try to explain what I'm doing here a bit better.

I am trying to tease apart the check for the existing workspace from the decision to backfill/create one because I want to make sure we _only_ backfill/create a workspace for valid snapshot requests.

This `Failure` block is what we expect _when we haven't created a workspace yet_

If the snapshot request isn't valid (eg: because it's for an Azure snapshot), it didn't seem right to have side effects and go create a workspace anyway, so I wanted to fail fast here in that case without chugging along and backfilling.

Does that make sense?

Reply:
Responding to this comment and the subesequent comment at once … what I believe the flow should be is:
1. Does the workspace exist _in Rawls_? If not, reject the request. This already happens inside `getV2WorkspaceContextAndPermissions()`, which is the first line of `SnapshotService.createSnapshot()`.
2. Is the snapshot-being-imported an Azure snapshot? If so, reject the request.
3. Is the workspace an Azure workspace? If so, reject the request.
4. Does the workspace exist _in WSM_? If not, create it.
5. Create the snapshot reference.

Specific to the `Success` or `Failure` of `aggregatedWorkspaceService.getAggregatedWorkspace()` - if the workspace is of type `WorkspaceType.RawlsWorkspace`, you'll get a `Success` as long as the workspace exists _in Rawls_, even if that workspace doesn't exist _in WSM_. So I don't think it is a reliable test for step 4 above.

Reply:
I hope this is in better shape now, see https://github.com/broadinstitute/rawls/pull/2556#discussion_r1346170905